Source/DafnyRuntime/DafnyRuntime.go

```diff
@@ -692,183 +692,412 @@ func (seq Seq) VerbatimString(asLiteral bool) string {
+  otherArray, ok := other.(*ArrayStruct)
+  if !ok {
+    return false
   }
+  return &_this.dims[0] == &otherArray.dims[0]
+}
+
+/***** ArrayForByte is the Array interface specialized for byte. *****/
+
+type ArrayForByte struct {
+  contents []byte // stored as a flat one-dimensional slice
+  dims     []int
+}
 
-  return sliceIsPrefixAfterLengthCheck(array.contents, array2.contents)
+func (_this ArrayForByte) dimensionCount() int {
+  return len(_this.dims)
 }
 
-// EqualsGeneric implements the EqualsGeneric interface.
-func (array *Array) EqualsGeneric(other interface{}) bool {
-  array2, ok := other.(*Array)
-  return ok && array.Equals(array2)
+func (_this ArrayForByte) dimensionLength(dim int) int {
+  return _this.dims[dim]
 }
 
-func (array *Array) findIndex(ixs ...int) int {
-  i := 0
-  size := 1
-  for d := len(array.dims) - 1; d >= 0; d-- {
-    i += size * ixs[d]
-    size *= array.dims[d]
-  }
-  return i
+func (_this ArrayForByte) ArrayGet1(index int) interface{} {
+  return _this.contents[index]
 }
 
-// Index gets the element at the given indices into the array.
-func (array *Array) Index(ixs ...Int) *interface{} {
-  if len(ixs) != len(array.dims) {
-    panic(fmt.Sprintf("Expected %d indices but got %d", len(array.dims), len(ixs)))
+func (_this ArrayForByte) ArraySet1(value interface{}, index int) {
+  _this.contents[index] = value.(byte)
+}
+
+func (_this ArrayForByte) ArrayGet1Byte(index int) byte {
+  return _this.contents[index]
+}
+
+func (_this ArrayForByte) ArraySet1Byte(value byte, index int) {
+  _this.contents[index] = value
+}
+
+func (_this ArrayForByte) ArrayGet1Char(index int) Char {
+  panic("cannot specialize general ArrayGet with char")
+}
+
+func (_this ArrayForByte) ArraySet1Char(value Char, index int) {
+  panic("cannot specialize general ArraySet with char")
+}
+
+func (_this ArrayForByte) anySlice(lo, hi Int) []interface{} {
+  if lo.IsNilInt() {
+    lo = Zero
   }
-  ints := make([]int, len(ixs))
-  for i, ix := range ixs {
-    ints[i] = ix.Int()
+  if hi.IsNilInt() {
+    hi = IntOf(len(_this.contents))
   }
-  return array.IndexInts(ints...)
+  iLo := lo.Int()
+  iHi := hi.Int()
+  anyArray := make([]interface{}, iHi - iLo)
+  for i := iLo; i < iHi; i++ {
+    anyArray[i] = _this.contents[i]
+  }
+  return anyArray
 }
 
-// IndexInts gets the element at the given indices into the array.
-func (array *Array) IndexInts(ixs ...int) *interface{} {
-  return &array.contents[array.findIndex(ixs...)]
+// ArrayForByte implements the EqualsGeneric interface.
+func (_this ArrayForByte) EqualsGeneric(other interface{}) bool {
+  otherArray, ok := other.(*ArrayForByte)
+  if !ok {
+    return false
+  }
+  return &_this.dims[0] == &otherArray.dims[0]
 }
 
-// Iterator iterates over the array.
-func (array *Array) Iterator() Iterator {
-  return sliceIterator(array.contents)
+/***** ArrayForChar is the Array interface specialized for char. *****/
+
+type ArrayForChar struct {
```

Comment:
Likewise we need an ArrayForCodePoint here too

Reply:
Done.